### PR TITLE
Only update needed columns when update user

### LINKED ISSUE #1882: User avatars seem to change color often
- Gitea version (or commit ref): 1.1.1+2-g2854c8a
- Git version: 2.12.2
- Operating system: Windows Server 2012
- Database (use `[x]`):
  - [x] MSSQL
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Not relevant because avatars use Gravatar.
- Log gist: Not relevant

User avatars seem to change color often. This is for auto-generated user avatars. I am in offline-mode and not using Gravatar.

Shapes in avatars remain the same. It seems I notice this almost daily. Is this a feature? 

I would prefer color to stay consistent.

### PR DESCRIPTION
**Pull request by lunny**
Also fix #1882. After this merged, I will back ported to branch release/v1.1.

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -877,6 +879,28 @@ func UpdateUser(u *User) error {
 	return updateUser(x, u)
 }
 
+// UpdateUserCols update user according special columns
+func UpdateUserCols(u *User, cols ...string) error {
+	// Organization does not need email
+	u.Email = strings.ToLower(u.Email)
```
Comment (sapk) [inline models/user.go:885]: Can't we iterate over cols array and with a switch only do needed operation when matching col ?

Example : https://play.golang.org/p/4pRMeP89Re

File: models/user.go [hunk 2]
```diff
@@ -877,6 +879,28 @@ func UpdateUser(u *User) error {
...
+			u.Avatar = base.HashEmail(u.AvatarEmail)
+		}
+	}
+
+	u.LowerName = strings.ToLower(u.Name)
+	u.Location = base.TruncateString(u.Location, 255)
+	u.Website = base.TruncateString(u.Website, 255)
+	u.Description = base.TruncateString(u.Description, 255)
+
+	_, err := x.Id(u.ID).Cols(cols...).Update(u)
```
Comment (ethantkoenig) [inline models/user.go]: The `updated_unix` column won't be updated. I think we need to do something similar to https://github.com/go-gitea/gitea/pull/2204

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by ethantkoenig** [comment]
Adding a new function seems like overkill to me, is there a reason we can't just fix the `updateUser(..)` function to not write to `u.Avatar` (and require the caller of `updateUser()` to update `u.Avatar` if they actually need it)?

**Comment by sapk** [comment]
@ethantkoenig We could  add a new func generateAvatar that is call before updateUser only when email is change or needed ?

**Comment by lunny** [comment]
@ethantkoenig I think `updateUser` is not corrected in many place where only need to update serval columns but update all the columns. So I added a new function `updateUserCols` to do that. And there are still some place where we should use `updateUser` to update all the columns.
We cannot finish the work only improve `updateUser` function, so I create the new one.
For fix of #1882, it's not the PR's main idea. #1882 's problem is when user login, the avatar will be updated, so the avatar changed. In fact, the only needed change column on login is `last_login_time` but it update all the columns. When I change to `updateUserCols`, the bug is also fixed.

File: models/user.go (hunk 2 above)
Comment (lunny) [reply models/user.go]: done.

File: models/user.go (hunk 1 above)
Comment (lunny) [reply models/user.go:885]: Didn't get what you mean.

**Review by ethantkoenig — APPROVED** [review]

**Comment by lunny** [comment]
@sapk that could be another PR.

**Comment by ethantkoenig** [comment]
LGTM

**Comment by lafriks** [comment]
LGTM

**Comment by lafriks** [comment]
Make LG-TM work

File: models/user.go (hunk 1 above)
Comment (sapk) [reply models/user.go:885]: Maybe it is overkill
```
// UpdateUserCols update user according special columns
func UpdateUserCols(u *User, cols ...string) error {
	//Format only when needed
	for _, c := range cols {
		switch c {
		case "email":
			// Organization does not need email
			u.Email = strings.ToLower(u.Email)
			if !u.IsOrganization() {
				if len(u.AvatarEmail) == 0 {
					u.AvatarEmail = u.Email
				}
			}
		case "name":
			u.LowerName = strings.ToLower(u.Name)
		case "location":
			u.Location = base.TruncateString(u.Location, 255)
		case "website":
			u.Website = base.TruncateString(u.Website, 255)
		case "description":
			u.Description = base.TruncateString(u.Description, 255)
		}
	}
	if !u.IsOrganization() {
		if len(u.AvatarEmail) > 0 {
			u.Avatar = base.HashEmail(u.AvatarEmail)
		}
	}
	_, err := x.Id(u.ID).Cols(cols...).Update(u)
	return err
}
```

**Pull request merged**